File: python/sort/base_sort.py
a = ["13572", "15734", "14269", "15384", "15721"]

def base_sort():
    item_len = len(a[0])

    for i in range(item_len-1, -1, -1):
        count_sort(i)

def count_sort(n):
    count_num = 10
    counting = [0] * count_num
    tmp = [None] * len(a)

    for i in range(0, len(a)):
        value = int(a[i][n])
        counting[value] += 1

    for i in range(1, count_num):
        counting[i] += counting[i-1]

    for i in range(len(a)-1, -1, -1):
        value = int(a[i][n])
        j = counting[value]
        tmp[j-1] = a[i]
        counting[value] -= 1

    # print(counting)
    # print(tmp)
    for i in range(0, len(a)):
        a[i] = tmp[i]


def main():
    base_sort()

    print(a)

if __name__ == '__main__':
    main()

File: c/sort/base_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define M 6
#define N 5

char a[M][N] = {"13572", "15734", "14269", "15384", "15721", "13852"};

void counting_sort(int n) {
    int counting[10] = {0,};
    char tmp[M][N] = {};
    
    for (int i = 0; i < M; i++) {
        char value = a[i][n] - 48;
        counting[value] += 1;
    }

    for (int i = 1; i < 10; i++) {
        counting[i] += counting[i-1];
    }

    for (int i = (M-1); i >= 0; i--) {
        char value = a[i][n] - 48;
        int j = counting[value];
        for (int k = 0; k < N; k++)
            tmp[j-1][k] = a[i][k];
        counting[value]--;
    }

    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            // printf("%c", tmp[i][j]);
            a[i][j] = tmp[i][j];
        }
        // printf("\n");
    }

}

void base_sort() {
    for (int i = (N-1); i >= 0; i--) {
        counting_sort(i);
    }
}

int main() {

    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            printf("%c", a[i][j]);
        }
        printf("\n");
    }
    printf("===================\n");

    base_sort();

    for (int i = 0; i < M; i++) {
        for (int j = 0; j < N; j++) {
            printf("%c", a[i][j]);
        }
        printf("\n");
    }

    return 0;
}